radixdlt-java/src/main/java/com/radixdlt/client/core/atoms/ParticleGroup.java

```diff
@@ -0,0 +1,141 @@
+@SerializerId2("PARTICLEGROUP")
+public class ParticleGroup extends SerializableObject {
+	@JsonProperty("particles")
+	@DsonOutput(DsonOutput.Output.ALL)
+	private final List<SpunParticle> particles = new ArrayList<>();
+
+	private ParticleGroup() {
+	}
+
+	public ParticleGroup(List<SpunParticle> particles) {
+		Objects.requireNonNull(particles, "particles is required");
+
+		this.particles.addAll(particles);
+	}
+
+	/**
+	 * Get a stream of the spun particles in this group
+	 */
+	public final Stream<SpunParticle> spunParticles() {
+		return this.particles.stream();
+	}
+
+	/**
+	 * Get a stream of particles of a certain spin in this group
+	 *
+	 * @param spin The spin to filter by
+	 * @return The particles in this group with that spin
+	 */
+	public final Stream<Particle> particles(Spin spin) {
+		return this.spunParticles().filter(p -> p.getSpin() == spin).map(SpunParticle::getParticle);
+	}
+
+	/**
+	 * Get a {@link ParticleGroup} with just one spun particle with the given spin and particle
+	 *
+	 * @param particle The particle
+	 * @param spin     The spin
+	 * @return The resulting particle group
+	 */
+	public static ParticleGroup just(Particle particle, Spin spin) {
+		return of(SpunParticle.of(particle, spin));
+	}
+
+	/**
+	 * Get a {@link ParticleGroup} consisting of the given particles
+	 */
+	public static ParticleGroup of(Iterable<SpunParticle> particles) {
+		Objects.requireNonNull(particles, "particles is required");
+
+		return new ParticleGroup(ImmutableList.copyOf(particles));
+	}
+
+	/**
+	 * Get a {@link ParticleGroup} consisting of the given particles
+	 */
+	public static ParticleGroup of(SpunParticle<?>... particles) {
+		Objects.requireNonNull(particles, "particles is required");
+
+		return new ParticleGroup(ImmutableList.copyOf(particles));
+	}
+
+	/**
+	 * Whether this {@link ParticleGroup} contains any particles
+	 */
+	public boolean hasParticles() {
+		return !this.particles.isEmpty();
+	}
+
+	/**
+	 * Get a build for a single {@link ParticleGroup}
+	 * @return The {@link ParticleGroupBuilder}
+	 */
+	public static ParticleGroupBuilder builder() {
+		return new ParticleGroupBuilder();
+	}
+
+	/**
+	 * A builder for immutable {@link ParticleGroup}s
+	 */
+	public static class ParticleGroupBuilder {
+		private List<SpunParticle> particles = new ArrayList<>();
+
+		private ParticleGroupBuilder() {
+		}
+
+		public final ParticleGroupBuilder addParticle(SpunParticle<?> spunParticle) {
+			Objects.requireNonNull(spunParticle, "spunParticle is required");
+
+			this.particles.add(spunParticle);
+
+			return this;
+		}
+
+		public final ParticleGroupBuilder addParticle(Particle particle, Spin spin) {
+			Objects.requireNonNull(particle, "particle is required");
+			Objects.requireNonNull(spin, "spin is required");
+
+			SpunParticle<?> spunParticle = SpunParticle.of(particle, spin);
+			this.particles.add(spunParticle);
+
+			return this;
+		}
+
+		public final boolean removeParticle(SpunParticle<?> spunParticle) {
```

Comment:
I don't think we want to expose ability to remove state from a builder as it would encourage bad uses and it doesn't look like this method is being used so I'd prefer to just remove it.

Reply:
Good catch, forgot to remove it after copying the code over from ParticleGroup when introducing builders.